lib/protocol/bep_fileinfo.go

```diff
@@ -1,56 +1,251 @@
-type FileIntf interface {
-	FileSize() int64
-	FileName() string
-	FileLocalFlags() uint32
-	IsDeleted() bool
-	IsInvalid() bool
-	IsIgnored() bool
-	IsUnsupported() bool
-	MustRescan() bool
-	IsReceiveOnlyChanged() bool
-	IsDirectory() bool
-	IsSymlink() bool
-	ShouldConflict() bool
-	HasPermissionBits() bool
-	SequenceNo() int64
-	BlockSize() int
-	FileVersion() Vector
-	FileType() FileInfoType
-	FilePermissions() uint32
-	FileModifiedBy() ShortID
-	ModTime() time.Time
-	PlatformData() PlatformData
-	InodeChangeTime() time.Time
-	FileBlocksHash() []byte
-}
-
-func (Hello) Magic() uint32 {
-	return HelloMessageMagic
+type FileInfo struct {
+	Name          string
+	Size          int64
+	ModifiedS     int64
+	ModifiedBy    ShortID
+	Version       Vector
+	Sequence      int64
+	Blocks        []BlockInfo
+	SymlinkTarget string
+	BlocksHash    []byte
+	Encrypted     []byte
+	Platform      PlatformData
+
+	Type         FileInfoType
+	Permissions  uint32
+	ModifiedNs   int32
+	RawBlockSize int32
+
+	// The local_flags fields stores flags that are relevant to the local
+	// host only. It is not part of the protocol, doesn't get sent or
+	// received (we make sure to zero it), nonetheless we need it on our
+	// struct and to be able to serialize it to/from the database.
+	LocalFlags uint32
+
+	// The version_hash is an implementation detail and not part of the wire
+	// format.
+	VersionHash []byte
+
+	// The time when the inode was last changed (i.e., permissions, xattrs
+	// etc changed). This is host-local, not sent over the wire.
+	InodeChangeNs int64
+
+	// The size of the data appended to the encrypted file on disk. This is
+	// host-local, not sent over the wire.
+	EncryptionTrailerSize int
+
+	Deleted       bool
+	RawInvalid    bool
+	NoPermissions bool
+
+	truncated bool // was created from a truncated file info without blocks
+}
+
+func (f *FileInfo) ToWire() *bep.FileInfo {
+	if f.truncated && !(f.IsDeleted() || f.IsInvalid() || f.IsIgnored()) {
+		panic("bug: must not serialize truncated file info")
+	}
+	blocks := make([]*bep.BlockInfo, len(f.Blocks))
+	for j, b := range f.Blocks {
+		blocks[j] = b.ToWire()
+	}
+	return &bep.FileInfo{
+		Name:          f.Name,
+		Size:          f.Size,
+		ModifiedS:     f.ModifiedS,
+		ModifiedBy:    uint64(f.ModifiedBy),
+		Version:       f.Version.ToWire(),
+		Sequence:      f.Sequence,
+		Blocks:        blocks,
+		SymlinkTarget: f.SymlinkTarget,
+		BlocksHash:    f.BlocksHash,
+		Encrypted:     f.Encrypted,
+		Type:          f.Type,
+		Permissions:   f.Permissions,
+		ModifiedNs:    f.ModifiedNs,
+		BlockSize:     f.RawBlockSize,
+		Platform:      f.Platform.toWire(),
+		Deleted:       f.Deleted,
+		Invalid:       f.RawInvalid,
+		NoPermissions: f.NoPermissions,
+	}
+}
+
+func (f *FileInfo) ToDB() *bep.FileInfo {
+	w := f.ToWire()
+	w.LocalFlags = f.LocalFlags
+	w.VersionHash = f.VersionHash
+	w.InodeChangeNs = f.InodeChangeNs
+	w.EncryptionTrailerSize = int32(f.EncryptionTrailerSize)
+	return w
+}
```

Comment:
I am a bit confused about this `ToWire` and `ToDB` distinction. It looks like it's about ignoring fields not part of BEP.  That seems sensible in principle, but imo there was no such distinction before this PR, we just always marshalled the entire thing (and calling code was responsible to (un)set fields as appropriate). I am not saying that was great, just that it worked. And I am not convinced by `ToDB` and `ToWire`, it feels like a very odd and hard to discover distinction. Until I stumbled over this, I always read `ToWire` as `ToProto` in my head - and for most types that's the case, a distinction only exists here. If we want to enforce the distinction, it seems safer and simpler to have two separate types, e.g. db.FileInfo vs protocol.FileInfo. Anyway, not this PR - for that I am hesitantly proposing to remove the `ToDB` and always convert all fields, maybe?

Reply:
I agree it's nice to enforce that in a single place/method. I think I still have a tiny preference for only `ToWire`, having this single instance of a separate method seems surprising to me. However fine to leave as is to me, whichever you prefer.

Reply:
How would you feel about a boolean argument `clearInternalFields` or something, to ToWire?

Reply:
Sounds good.

(I actually didn't want to just send this one comment yesterday in a half-assed followup, but finish up with everything here - just as usual got distracted and then forgot about it until now...)